File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    answer = [0 for _ in range(len(prices))]
    stack = []
    
    for i in range(len(prices)-1, -1, -1):
        # 거꾸로 탐색
        while stack and prices[i] <= prices[stack[-1]]:
            ## 가격 떨어지지 않음
            stack.pop()
            
        if stack:
            answer[i] = stack[-1]-i
        else:
            answer[i] = len(prices)-1-i
            
        stack.append(i)
        
    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int[prices.length];
        
        Deque<Integer> stack = new ArrayDeque<>();
        
        for(int i=prices.length-1; i>=0; --i){
            while(!stack.isEmpty() && prices[i] <= prices[stack.peekLast()]){
                stack.pollLast();
            }
            
            if (!stack.isEmpty()){
                answer[i] = stack.peekLast()-i;
            } else {
                answer[i] = prices.length-1-i;
            }
            
            stack.offerLast(i);
        }
        
        
        return answer;
    }
}